src/transformers/optimization.py

```diff
@@ -387,6 +387,84 @@ def get_cosine_with_min_lr_schedule_with_warmup(
     return LambdaLR(optimizer, lr_lambda, last_epoch)
 
 
+def _get_cosine_with_min_lr_schedule_with_warmup_lr_rate_lambda(
+    current_step: int,
+    *,
+    num_warmup_steps: int,
+    num_training_steps: int,
+    num_cycles: float,
+    min_lr_rate: float = 0.0,
+    warmup_lr_rate: float = None,
+):
+    if current_step < num_warmup_steps:
+        if warmup_lr_rate is None:
+            return float(current_step + 1) / float(max(1, num_warmup_steps))
+        else:
+            return float(warmup_lr_rate) + float(1.0 - warmup_lr_rate) * float(current_step) / float(
+                max(1, num_warmup_steps - 1)
+            )
+    progress = float(current_step - num_warmup_steps + 1) / float(max(1, num_training_steps - num_warmup_steps))
```

Comment:
Similarly here - we could add input validation to ensure that num_training_steps > num_warmup_steps instead of needing this `max` fallback